**Problem 2.1**

In [1]:
import numpy as np
import pandas as pd

from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

In [2]:
df = pd.read_excel('/Users/kdiepholz/Downloads/Online Retail.xlsx')

In [3]:
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

In [4]:
df.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

In [12]:
Basket = (df[df['Country']=="France"]
          .groupby(['InvoiceNo', 'Description'])['Quantity']
          .sum().unstack().reset_index().fillna(0)
          .set_index('InvoiceNo'))

Basket.head()

Description   50'S CHRISTMAS GIFT BAG LARGE   DOLLY GIRL BEAKER  \
InvoiceNo                                                         
536370                                  0.0                 0.0   
536852                                  0.0                 0.0   
536974                                  0.0                 0.0   
537065                                  0.0                 0.0   
537463                                  0.0                 0.0   

Description   I LOVE LONDON MINI BACKPACK   NINE DRAWER OFFICE TIDY  \
InvoiceNo                                                             
536370                                0.0                       0.0   
536852                                0.0                       0.0   
536974                                0.0                       0.0   
537065                                0.0                       0.0   
537463                                0.0                       0.0   

Description   SET 2 TEA TOWELS I LOVE LONDON    SPACEBOY BABY GIFT SET  \
InvoiceNo                                                                
536370                                   24.0                      0.0   
536852                                    0.0                      0.0   
536974                                    0.0                      0.0   
537065                                    0.0                      0.0   
537463                                    0.0                      0.0   

Description   TRELLIS COAT RACK  10 COLOUR SPACEBOY PEN  \
InvoiceNo                                                 
536370                      0.0                     0.0   
536852                      0.0                     0.0   
536974                      0.0                     0.0   
537065                      0.0                     0.0   
537463                      0.0                     0.0   

Description  12 COLOURED PARTY BALLOONS  12 EGG HOUSE PAINTED WOOD  ...  \
InvoiceNo                                                           ...   
536370                              0.0                        0.0  ...   
536852                              0.0                        0.0  ...   
536974                              0.0                        0.0  ...   
537065                              0.0                        0.0  ...   
537463                              0.0                        0.0  ...   

Description  WRAP VINTAGE PETALS  DESIGN  YELLOW COAT RACK PARIS FASHION  \
InvoiceNo                                                                  
536370                               0.0                             0.0   
536852                               0.0                             0.0   
536974                               0.0                             0.0   
537065                               0.0                             0.0   
537463                               0.0                             0.0   

Description  YELLOW GIANT GARDEN THERMOMETER  YELLOW SHARK HELICOPTER  \
InvoiceNo                                                               
536370                                   0.0                      0.0   
536852                                   0.0                      0.0   
536974                                   0.0                      0.0   
537065                                   0.0                      0.0   
537463                                   0.0                      0.0   

Description  ZINC  STAR T-LIGHT HOLDER   ZINC FOLKART SLEIGH BELLS  \
InvoiceNo                                                            
536370                              0.0                        0.0   
536852                              0.0                        0.0   
536974                              0.0                        0.0   
537065                              0.0                        0.0   
537463                              0.0                        0.0   

Description  ZINC HERB GARDEN CONTAIN

In [13]:
def sum_to_boolean(x):
    if x<=0:
        return 0
    else:
        return 1

Basket_Final = Basket.applymap(sum_to_boolean)

In [14]:
Basket_Final.drop('POSTAGE', inplace=True, axis=1)

Basket_Final.head()

Description   50'S CHRISTMAS GIFT BAG LARGE   DOLLY GIRL BEAKER  \
InvoiceNo                                                         
536370                                    0                   0   
536852                                    0                   0   
536974                                    0                   0   
537065                                    0                   0   
537463                                    0                   0   

Description   I LOVE LONDON MINI BACKPACK   NINE DRAWER OFFICE TIDY  \
InvoiceNo                                                             
536370                                  0                         0   
536852                                  0                         0   
536974                                  0                         0   
537065                                  0                         0   
537463                                  0                         0   

Description   SET 2 TEA TOWELS I LOVE LONDON    SPACEBOY BABY GIFT SET  \
InvoiceNo                                                                
536370                                      1                        0   
536852                                      0                        0   
536974                                      0                        0   
537065                                      0                        0   
537463                                      0                        0   

Description   TRELLIS COAT RACK  10 COLOUR SPACEBOY PEN  \
InvoiceNo                                                 
536370                        0                       0   
536852                        0                       0   
536974                        0                       0   
537065                        0                       0   
537463                        0                       0   

Description  12 COLOURED PARTY BALLOONS  12 EGG HOUSE PAINTED WOOD  ...  \
InvoiceNo                                                           ...   
536370                                0                          0  ...   
536852                                0                          0  ...   
536974                                0                          0  ...   
537065                                0                          0  ...   
537463                                0                          0  ...   

Description  WRAP VINTAGE PETALS  DESIGN  YELLOW COAT RACK PARIS FASHION  \
InvoiceNo                                                                  
536370                                 0                               0   
536852                                 0                               0   
536974                                 0                               0   
537065                                 0                               0   
537463                                 0                               0   

Description  YELLOW GIANT GARDEN THERMOMETER  YELLOW SHARK HELICOPTER  \
InvoiceNo                                                               
536370                                     0                        0   
536852                                     0                        0   
536974                                     0                        0   
537065                                     0                        0   
537463                                     0                        0   

Description  ZINC  STAR T-LIGHT HOLDER   ZINC FOLKART SLEIGH BELLS  \
InvoiceNo                                                            
536370                                0                          0   
536852                                0                          0   
536974                                0                          0   
537065                                0                          0   
537463                                0                          0   

Description  ZINC HERB GARDEN CONTAIN

In [15]:
Frequent_itemsets = apriori(Basket_Final, min_support = 0.06, use_colnames = True)

Frequent_itemsets.sort_values('support', ascending = False).head()

support                               itemsets
22  0.160521                   (RABBIT NIGHT LIGHT)
26  0.154013        (RED TOADSTOOL LED NIGHT LIGHT)
21  0.145336     (PLASTERS IN TIN WOODLAND ANIMALS)
18  0.143167       (PLASTERS IN TIN CIRCUS PARADE )
30  0.134490  (ROUND SNACK BOXES SET OF4 WOODLAND )

Using the apriori algorithm, we can see that the (RABBIT NIGHT LIGHT) has the highest support, with .160521

In [16]:
Frequent_itemsets = apriori(Basket_Final, min_support = 0.05, use_colnames = True)

Frequent_itemsets.sort_values('support', ascending = False).head()

support                               itemsets
34  0.160521                   (RABBIT NIGHT LIGHT)
40  0.154013        (RED TOADSTOOL LED NIGHT LIGHT)
32  0.145336     (PLASTERS IN TIN WOODLAND ANIMALS)
29  0.143167       (PLASTERS IN TIN CIRCUS PARADE )
45  0.134490  (ROUND SNACK BOXES SET OF4 WOODLAND )

We set the minimum support threshhold to 5%, or .05 as min_support

Using this, we can run the association_rules module with our updated frequent itemsets

In [23]:
Asso_Rules = association_rules(Frequent_itemsets, metric = "lift", min_threshold = 8)
Asso_Rules.sort_values('lift',ascending = False)

antecedents  \
6                     (CHILDRENS CUTLERY DOLLY GIRL )   
7                       (CHILDRENS CUTLERY SPACEBOY )   
11  (ALARM CLOCK BAKELIKE GREEN, ALARM CLOCK BAKEL...   
14                        (ALARM CLOCK BAKELIKE RED )   
13                       (ALARM CLOCK BAKELIKE GREEN)   
12  (ALARM CLOCK BAKELIKE RED , ALARM CLOCK BAKELI...   
2                        (ALARM CLOCK BAKELIKE GREEN)   
3                         (ALARM CLOCK BAKELIKE RED )   
10  (ALARM CLOCK BAKELIKE GREEN, ALARM CLOCK BAKEL...   
15                        (ALARM CLOCK BAKELIKE PINK)   
4                         (ALARM CLOCK BAKELIKE RED )   
5                         (ALARM CLOCK BAKELIKE PINK)   
16  (SET/6 RED SPOTTY PAPER CUPS, SET/20 RED RETRO...   
19                    (SET/6 RED SPOTTY PAPER PLATES)   
1                         (ALARM CLOCK BAKELIKE PINK)   
0                        (ALARM CLOCK BAKELIKE GREEN)   
17  (SET/20 RED RETROSPOT PAPER NAPKINS , SET/6 RE...   
18                      (SET/6 RED SPOTTY PAPER CUPS)   
9                     (SET/6 RED SPOTTY PAPER PLATES)   
8                       (SET/6 RED SPOTTY PAPER CUPS)   

                                          consequents  antecedent support  \
6                       (CHILDRENS CUTLERY SPACEBOY )            0.060738   
7                     (CHILDRENS CUTLERY DOLLY GIRL )            0.058568   
11                        (ALARM CLOCK BAKELIKE RED )            0.062907   
14  (ALARM CLOCK BAKELIKE GREEN, ALARM CLOCK BAKEL...            0.080260   
13  (ALARM CLOCK BAKELIKE RED , ALARM CLOCK BAKELI...            0.082430   
12                       (ALARM CLOCK BAKELIKE GREEN)            0.062907   
2                         (ALARM CLOCK BAKELIKE RED )            0.082430   
3                        (ALARM CLOCK BAKELIKE GREEN)            0.080260   
10                        (ALARM CLOCK BAKELIKE PINK)            0.067245   
15  (ALARM CLOCK BAKELIKE GREEN, ALARM CLOCK BAKEL...            0.086768   
4                         (ALARM CLOCK BAKELIKE PINK)            0.080260   
5                         (ALARM CLOCK BAKELIKE RED )            0.086768   
16                    (SET/6 RED SPOTTY PAPER PLATES)            0.086768   
19  (SET/6 RED SPOTTY PAPER CUPS, SET/20 RED RETRO...            0.108460   
1                        (ALARM CLOCK BAKELIKE GREEN)            0.086768   
0                         (ALARM CLOCK BAKELIKE PINK)            0.082430   
17                      (SET/6 RED SPOTTY PAPER CUPS)            0.086768   
18  (SET/20 RED RETROSPOT PAPER NAPKINS , SET/6 RE...            0.117137   
9                       (SET/6 RED SPOTTY PAPER CUPS)            0.108460   
8                     (SET/6 RED SPOTTY PAPER PLATES)            0.117137   

    consequent support   support  confidence       lift  leverage  conviction  
6             0.058568  0.054230    0.892857  15.244709  0.050673    8.786696  
7             0.060738  0.054230    0.925926  15.244709  0.050673   12.680043  
11            0.080260  0.054230    0.862069  10.740913  0.049181    6.668113  
14            0.062907  0.054230    0.675676  10.740913  0.049181    2.889371  
13            0.062907  0.054230    0.657895  10.458258  0.049045    2.739196  
12            0.082430  0.054230    0.862069  10.458258  0.049045    6.652386  
2             0.080260  0.067245    0.815789  10.164296  0.060629    4.992873  
3             0.082430  0.067245    0.837838  10.164296  0.060629    5.658351  
10            0.086768  0.054230    0.806452   9.294355  0.048395    4.718366  
15            0.067245  0.054230    0.625000   9.294355  0.048395    2.487346  
4             0.086768  0.062907    0.783784   9.033108  0.055943    4.223698  
5             0.080260  0.062907    0.725000   9.033108  0.055943    3.344508  
16            0.108460  0.084599    0.975000   8.989500  0.075188   35.661605  
19            0.086768  0.084599    0.780000   8.989500  0.075188    4.151055  
1             0.082430  0.

From this we can see that **Childrens cutlery spaceboy and childrens cutlery dolly girl** are purchased together. 
We can also see that **alarm clock bakelike green and alarm clock bakelike red** are purchased together. 
Finally we can see that **Set/6 red spotty paper cups and set/6 red spotty paper plates** are purchased together. 

In [21]:
Asso_Rules = association_rules(Frequent_itemsets, metric = "confidence", min_threshold = .8)
Asso_Rules.sort_values('confidence',ascending = False)

antecedents  \
9   (SET/6 RED SPOTTY PAPER CUPS, SET/20 RED RETRO...   
11  (SET/20 RED RETROSPOT PAPER NAPKINS , SET/6 RE...   
5                     (SET/6 RED SPOTTY PAPER PLATES)   
3                       (CHILDRENS CUTLERY SPACEBOY )   
2                     (CHILDRENS CUTLERY DOLLY GIRL )   
4                       (SET/6 RED SPOTTY PAPER CUPS)   
7   (ALARM CLOCK BAKELIKE GREEN, ALARM CLOCK BAKEL...   
8   (ALARM CLOCK BAKELIKE RED , ALARM CLOCK BAKELI...   
1                         (ALARM CLOCK BAKELIKE RED )   
0                        (ALARM CLOCK BAKELIKE GREEN)   
10  (SET/6 RED SPOTTY PAPER CUPS, SET/6 RED SPOTTY...   
6   (ALARM CLOCK BAKELIKE GREEN, ALARM CLOCK BAKEL...   

                              consequents  antecedent support  \
9         (SET/6 RED SPOTTY PAPER PLATES)            0.086768   
11          (SET/6 RED SPOTTY PAPER CUPS)            0.086768   
5           (SET/6 RED SPOTTY PAPER CUPS)            0.108460   
3         (CHILDRENS CUTLERY DOLLY GIRL )            0.058568   
2           (CHILDRENS CUTLERY SPACEBOY )            0.060738   
4         (SET/6 RED SPOTTY PAPER PLATES)            0.117137   
7             (ALARM CLOCK BAKELIKE RED )            0.062907   
8            (ALARM CLOCK BAKELIKE GREEN)            0.062907   
1            (ALARM CLOCK BAKELIKE GREEN)            0.080260   
0             (ALARM CLOCK BAKELIKE RED )            0.082430   
10  (SET/20 RED RETROSPOT PAPER NAPKINS )            0.104121   
6             (ALARM CLOCK BAKELIKE PINK)            0.067245   

    consequent support   support  confidence       lift  leverage  conviction  
9             0.108460  0.084599    0.975000   8.989500  0.075188   35.661605  
11            0.117137  0.084599    0.975000   8.323611  0.074435   35.314534  
5             0.117137  0.104121    0.960000   8.195556  0.091417   22.071584  
3             0.060738  0.054230    0.925926  15.244709  0.050673   12.680043  
2             0.058568  0.054230    0.892857  15.244709  0.050673    8.786696  
4             0.108460  0.104121    0.888889   8.195556  0.091417    8.023861  
7             0.080260  0.054230    0.862069  10.740913  0.049181    6.668113  
8             0.082430  0.054230    0.862069  10.458258  0.049045    6.652386  
1             0.082430  0.067245    0.837838  10.164296  0.060629    5.658351  
0             0.080260  0.067245    0.815789  10.164296  0.060629    4.992873  
10            0.112798  0.084599    0.812500   7.203125  0.072854    4.731743  
6             0.086768  0.054230    0.806452   9.294355  0.048395    4.718366

From our confidence association rules, we can see that **set/6 red spotty paper cups and set/6 red spotty paper plates** and **set/20 red retrospot paper napkins and set/6 red spotty paper cups** and **childrens cutlery spaceboy and childrens cutlery dolly girl** are all purchased together.

The rule with the highest lift is not the same as the rule with the highest confidence. This could be due to a number of reasons. One of the largest reasons is because lift and confidence take on a different equation. In addition, they take into account different things, with lift being the highest number of times something was purchased together and confidence being the ratio between how many times it was purchased together vs the number of times it was purchased in general.

**Problem 2.2**

In [4]:
import math
import numpy as np
import pandas as pd

from sklearn.metrics.pairwise import cosine_similarity

In [5]:
data = pd.read_csv('75000-out2-binary.csv')

In [6]:
data.head()

Transaction Number  Chocolate Cake  Lemon Cake  Casino Cake  Opera Cake  \
0                   1               0           0            0           0   
1                   2               0           0            0           0   
2                   3               0           0            0           1   
3                   4               0           0            0           0   
4                   5               0           0            0           0   

   Strawberry Cake  Truffle Cake  Chocolate Eclair  Coffee Eclair  \
0                0             0                 0              0   
1                0             0                 0              1   
2                0             0                 0              0   
3                0             1                 0              0   
4                0             0                 1              0   

   Vanilla Eclair  ...  Lemon Lemonade  Raspberry Lemonade  Orange Juice  \
0               0  ...               0                   0             0   
1               0  ...               0                   0             0   
2               0  ...               0                   0             1   
3               0  ...               0                   0             0   
4               0  ...               0                   0             1   

   Green Tea  Bottled Water  Hot Coffee  Chocolate Coffee  \
0          0              0           0                 0   
1          0              0           1                 0   
2          0              0           0                 0   
3          0              0           0                 0   
4          0              0           0                 0   

   Vanilla Frappuccino  Cherry Soda  Single Espresso  
0                    0            0                0  
1                    0            0                0  
2                    0            0                0  
3                    1            0                0  
4                    0            0                0  

[5 rows x 51 columns]

In [7]:
data.describe()

Transaction Number  Chocolate Cake    Lemon Cake   Casino Cake  \
count        75000.000000    75000.000000  75000.000000  75000.000000   
mean         37500.500000        0.083533      0.083613      0.075013   
std          21650.779432        0.276689      0.276809      0.263415   
min              1.000000        0.000000      0.000000      0.000000   
25%          18750.750000        0.000000      0.000000      0.000000   
50%          37500.500000        0.000000      0.000000      0.000000   
75%          56250.250000        0.000000      0.000000      0.000000   
max          75000.000000        1.000000      1.000000      1.000000   

         Opera Cake  Strawberry Cake  Truffle Cake  Chocolate Eclair  \
count  75000.000000     75000.000000  75000.000000      75000.000000   
mean       0.082093         0.092640      0.082240          0.042373   
std        0.274509         0.289929      0.274732          0.201441   
min        0.000000         0.000000      0.000000          0.000000   
25%        0.000000         0.000000      0.000000          0.000000   
50%        0.000000         0.000000      0.000000          0.000000   
75%        0.000000         0.000000      0.000000          0.000000   
max        1.000000         1.000000      1.000000          1.000000   

       Coffee Eclair  Vanilla Eclair  ...  Lemon Lemonade  Raspberry Lemonade  \
count   75000.000000    75000.000000  ...    75000.000000        75000.000000   
mean        0.109240        0.042520  ...        0.068240            0.067747   
std         0.311942        0.201774  ...        0.252159            0.251312   
min         0.000000        0.000000  ...        0.000000            0.000000   
25%         0.000000        0.000000  ...        0.000000            0.000000   
50%         0.000000        0.000000  ...        0.000000            0.000000   
75%         0.000000        0.000000  ...        0.000000            0.000000   
max         1.000000        1.000000  ...        1.000000            1.000000   

       Orange Juice     Green Tea  Bottled Water    Hot Coffee  \
count  75000.000000  75000.000000   75000.000000  75000.000000   
mean       0.091613      0.062467       0.075200      0.102667   
std        0.288481      0.242003       0.263716      0.303525   
min        0.000000      0.000000       0.000000      0.000000   
25%        0.000000      0.000000       0.000000      0.000000   
50%        0.000000      0.000000       0.000000      0.000000   
75%        0.000000      0.000000       0.000000      0.000000   
max        1.000000      1.000000       1.000000      1.000000   

       Chocolate Coffee  Vanilla Frappuccino   Cherry Soda  Single Espresso  
count      75000.000000         75000.000000  75000.000000     75000.000000  
mean           0.083147             0.077467      0.061987         0.067973  
std            0.276106             0.267332      0.241133         0.251702  
min            0.000000             0.000000      0.000000         0.000000  
25%            0.000000             0.000000      0.000000         0.000000  
50%            0.000000             0.000000      0.000000         0.000000  
75%            0.000000             0.000000      0.000000         0.000000  
max            1.000000             1.000000      1.000000         1.000000  

[8 rows x 51 columns]

In [8]:
item1_name = 'Chocolate Coffee'
item2_name = 'Chocolate Cake'

selection = data[[item1_name,
                  item2_name]]

In [9]:
item1_count = selection[item1_name] == 1
item2_count = selection[item2_name] == 1

In [10]:
f = selection.groupby([item1_count,
                       item2_count]).count()
print(f)

                                 Chocolate Coffee  Chocolate Cake
Chocolate Coffee Chocolate Cake                                  
False            False                      65802           65802
                 True                        2962            2962
True             False                       2933            2933
                 True                        3303            3303


In [11]:
f00 = f[item1_name][0][0]
f01 = f[item1_name][0][1]
f10 = f[item2_name][1][0]
f11 = f[item2_name][1][1]

print(f00)
print(f01)
print(f10)
print(f11)

65802
2962
2933
3303


In order to calculate false positives and false negatives, we can add them together

In [12]:
f1p = f11 + f10
fp1 = f11 + f01
f0p = f01 + f00
fp0 = f10 + f00

print(f1p)
print(fp1)
print(f0p)
print(fp0)

6236
6265
68764
68735


In [13]:
n = f1p + f0p
print(n)

n = fp1 + fp0
print(n)

75000
75000


True positive over all possible (75k)

In [17]:
stren = f11/n
print(s)
conf = f11/f1p
print(conf)

0.04404
0.5296664528543938


In [20]:
stren_a = f1p / n
stren_b = fp1 / n
stren_ab = f11 / n

conf_ab = stren_ab / stren_a
conf_ba = stren_ab / stren_b

I = stren_ab / (math.sqrt(stren_a * stren_b))
print(I)

0.5284391468918322


We can see that the binary coorelation coefficent is .5284

To see if they are the same with Chocolate cake -> Chocolate coffee, we just flip the items

In [21]:
item1_name = 'Chocolate Cake'
item2_name = 'Chocolate Coffee'

selection = data[[item1_name,
                  item2_name]]

In [22]:
item1_count = selection[item1_name] == 1
item2_count = selection[item2_name] == 1

In [23]:
f = selection.groupby([item1_count,
                       item2_count]).count()
print(f)

                                 Chocolate Cake  Chocolate Coffee
Chocolate Cake Chocolate Coffee                                  
False          False                      65802             65802
               True                        2933              2933
True           False                       2962              2962
               True                        3303              3303


Looking at these values we can conclude that they are the same either way that the itemsets are put